### PR TITLE
ref: Rewrite the native SDK in C

### LINKED ISSUE #29: add install action to premake
it would be awesome to be able to pass an install prefix to premake and have all the includes, libs and binaries installed there 

https://github.com/premake/premake-core/wiki/Command-Line-Arguments#creating-new-actions

### LINKED ISSUE #60: Add compatibility with android-16 for NDK
I had to bump android to 21 for now to make some progress. We should look into what we need to do to support android-16.

refs #59

### LINKED ISSUE #113: Compile as static library ?
There is no way to integrate this as static lib ? I compiled with /MT as well (with VS) but my application still asks for sentry.dll.

This is **very inconvenient**, especially due to the nature of my software. 

A solution would be appreciated.

Side notes: 

- Why is the code so ugly ?! I mean... it's awful. I`m not expecting it to be perfect, but it looks like it was written in a rush...
- Why is there so much C and C++ combined ?...
- Too many strings in plain, for security apps this is a risk and concern...
- Plenty more to say really...

Anyway, a solution to build as static lib is appreciated at this point; the rest... no further comments.

Thanks

### LINKED ISSUE #114: CMake support for all platform？
I notice there is a CMakeLists.txt in project, but it seems only support for android.

### PR DESCRIPTION
**Pull request by Swatinem**
A complete rewrite of the `sentry-native` SDK in C, very much WIP at this point.

fixes #60 
fixes #113 
fixes #114 

closes #29

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,29 +1,136 @@
...
+	${SENTRY_NATIVE_ALL_SOURCES}
+	${SENTRY_NATIVE_TEST_SOURCES}
+)
+target_compile_definitions("sentry_tests" PUBLIC -DSENTRY_UNITTEST)
+target_link_libraries("sentry_tests" ${LINK_LIBRARIES})
+
+# to fix some issues with tests (dladdr can't find functions otherwise)
+if(CMAKE_SYSTEM_NAME STREQUAL "Linux" OR CMAKE_SYSTEM_NAME STREQUAL "Android")
+	target_link_libraries("sentry_tests" "-Wl,--build-id=uuid,-E")
+endif()
```
Comment (marandaneto) [inline CMakeLists.txt]: this is causing issues on Android if building on Windows.

https://android.googlesource.com/platform/ndk/+/master/docs/BuildSystemMaintainers.md#additional-required-arguments
`uuid` is not the default form of `--build-id`, why do we need this one and not the default one?

What's about `LLD`? we could try `-fuse-ld=lld`, it's better at Windows, bfd/gold are very Linux first.

Also, not sure if a fix should come from sentry-native or sentry-android-ndk, as build was working before and now it doesn't.

Just excluding `sentry_tests` on sentry-android-ndk seems wrong to me, I'd need to do multiple times probably.

@Swatinem

Comment (mitsuhiko) [reply CMakeLists.txt]: Sentry is good with any build id for as long as there is one. We used to see SHA1 or other checksums in the field.

Comment (marandaneto) [reply CMakeLists.txt]: @Swatinem so unless you have a special reason to require `uuid` type, would it be ok to use the standard one?

Comment (jan-auer) [reply CMakeLists.txt]: uuid is just the fastest, feel free to replace it.

Comment (jan-auer) [reply CMakeLists.txt]: The important bit is -E

Comment (marandaneto) [reply CMakeLists.txt]: @jan-auer I see, thanks.
https://github.com/getsentry/sentry-native/pull/134

**Comment by Amphaal** [comment]
GCC (from GNU 9.2.0) throws a bunch of "incompatible-pointer-types" warnings while compiling. Are those fixable ?

**Comment by Swatinem** [comment]
I noticed that we have certain compiler warnings, which are very dependant on the compiler version.
I would like to clean those up, and add more test configurations for all the platforms we want to support. But that will still take some time.

**Pull request merged**